display_list/dl_color.h

```diff
@@ -58,58 +77,99 @@ struct DlColor {
   static constexpr DlColor kCornflowerBlue()     {return DlColor(0xFF6495ED);};
   // clang-format on
 
-  constexpr bool isOpaque() const { return getAlpha() == 0xFF; }
-  constexpr bool isTransparent() const { return getAlpha() == 0; }
+  constexpr bool isOpaque() const { return alpha_ >= 1.f; }
+  constexpr bool isTransparent() const { return alpha_ <= 0.f; }
+
+  constexpr int getAlpha() const { return toC(alpha_); }
+  constexpr int getRed() const { return toC(red_); }
+  constexpr int getGreen() const { return toC(green_); }
+  constexpr int getBlue() const { return toC(blue_); }
 
-  constexpr int getAlpha() const { return argb_ >> 24; }
-  constexpr int getRed() const { return (argb_ >> 16) & 0xFF; }
-  constexpr int getGreen() const { return (argb_ >> 8) & 0xFF; }
-  constexpr int getBlue() const { return argb_ & 0xFF; }
+  constexpr DlScalar getAlphaF() const { return alpha_; }
+  constexpr DlScalar getRedF() const { return red_; }
+  constexpr DlScalar getGreenF() const { return green_; }
+  constexpr DlScalar getBlueF() const { return blue_; }
 
-  constexpr DlScalar getAlphaF() const { return toF(getAlpha()); }
-  constexpr DlScalar getRedF() const { return toF(getRed()); }
-  constexpr DlScalar getGreenF() const { return toF(getGreen()); }
-  constexpr DlScalar getBlueF() const { return toF(getBlue()); }
+  constexpr DlColorSpace getColorSpace() const { return color_space_; }
 
   constexpr uint32_t premultipliedArgb() const {
     if (isOpaque()) {
-      return argb_;
+      return argb();
     }
     DlScalar f = getAlphaF();
-    return (argb_ & 0xFF000000) |       //
+    return (argb() & 0xFF000000) |      //
            toC(getRedF() * f) << 16 |   //
            toC(getGreenF() * f) << 8 |  //
            toC(getBlueF() * f);
   }
 
   constexpr DlColor withAlpha(uint8_t alpha) const {  //
-    return DlColor((argb_ & 0x00FFFFFF) | (alpha << 24));
+    return DlColor((argb() & 0x00FFFFFF) | (alpha << 24));
```

Comment:
I would convert the alpha to float and then use the constructor that takes the raw r_, g_, b_ values (and colorspace) rather than convert to uint32, apply the alpha, and convert back?

Same comment for withR/G/B methods.

Reply:
Basically `return withFoo(ToF(arg))`?

Reply:
Also, modulateOpacity() could be more direct in this way. It's basically withAlphaF before withAlphaF was created in this patch.

Reply:
I think if people are using these old `with*` setters they want to have the old behavior, which is clamping the numbers to 255.  `modulateOpacity` we should consider updating.  I would want to do that in a follow up PR though to avoid anyone that is relying on this behavior.